### ISSUE TITLE
[v3.6] Crash after hot update 

### ISSUE THREAD

**Issue by dumganhar** [cocos/cocos-engine#11339]
It's dangerous to use a thread that is detached.

<img width="713" alt="WeChatd045b816d0efec0f19686a16d4572108" src="https://user-images.githubusercontent.com/493372/171546930-e577ba10-7846-4d14-9844-b6a29678b321.png">
<img width="1972" alt="WeChatb2483f91871ca867ebd70bf33ae4e764" src="https://user-images.githubusercontent.com/493372/171546945-e9073a1a-0bb5-4697-b10c-e7958d745aa3.png">

<img width="683" alt="WeChatc9d23b370d6c5540a407838ee7b08dd2" src="https://user-images.githubusercontent.com/493372/171549336-838fa972-c995-4a6f-8cad-e4c62ad480bf.png">

<img width="935" alt="WeChat456acf2d194dda4563e9be21df6715e2" src="https://user-images.githubusercontent.com/493372/171549348-48377ab8-22d4-44af-a452-5f1b84cce5ed.png">

## Step of code execute flow, read the `// -->` comments

```c++
void MessageQueue::runConsumerThread() noexcept {
    if (_immediateMode || _workerAttached) return;

    _reader.terminateConsumerThread = false;
    _reader.flushingFinished = false;

    std::thread consumerThread(&MessageQueue::consumerThreadLoop, this);
    consumerThread.detach(); // --> 1. consumerThreadLoop is a detached thread
    _workerAttached = true;
}

void DeviceAgent::doDestroy() {
    ......

    // --> 2. terminateConsumerThread in main thread
    _mainMessageQueue->terminateConsumerThread();

    _mainMessageQueue->~MessageQueue();
    CC_FREE_ALIGN(_mainMessageQueue);// --> 9. Delete MessageQueue instance, release its memory
    _mainMessageQueue = nullptr;
}

void MessageQueue::terminateConsumerThread() noexcept {
    if (_immediateMode || !_workerAttached) return;

    EventSem event;
    EventSem *const pEvent = &event;

    ReaderContext *const pR = &_reader;

    ccnew_placement(allocate<TerminateConsumerThreadMessage>(1)) TerminateConsumerThreadMessage(pEvent, pR); // --> 3. Send a terminate message to queue

    kick(); // --> 4. Tell the consumer thread to work
    event.wait(); // --> 5. Wait the last TerminateConsumerThreadMessage to be executed
    // --> 8. terminateConsumerThread done
}

void MessageQueue::consumerThreadLoop() noexcept {
    while (!_reader.terminateConsumerThread) {// --> 10. Bang~~~, MessageQueue was freed in step 9. Here checks the freed memory. Crash is found by Clang Address Sanitizer
        AutoReleasePool autoReleasePool;
        flushMessages(); // --> 6. flushMessages all messages in queue, handle the last terminate message
        // --> 8. continue to while loop
    }

    _workerAttached = false;// --> 11. Bang~~~, Here checks the freed memory second time. Crash is found by Clang Address Sanitizer
}

void TerminateConsumerThreadMessage::execute() noexcept {
    _reader->terminateConsumerThread = true;
    _reader->flushingFinished = true;
    _event->signal(); // --> 7. Notify main thread the last message was done, it just set two flags in reader.
}

```

**Comment by dumganhar** [comment]
https://github.com/cocos/engine-native/pull/4008

**Comment by dumganhar** [comment]
## How to Fix

Don't detach consumer thread, save the instance and join it at the end of terminateConsumerThread.

```c++
void MessageQueue::runConsumerThread() noexcept {
    if (_immediateMode || _workerAttached) return;

    _reader.terminateConsumerThread = false;
    _reader.flushingFinished = false;

    _consumerThread = ccnew std::thread(&MessageQueue::consumerThreadLoop, this);
    // Don't detach thread
    _workerAttached = true;
}

void MessageQueue::terminateConsumerThread() noexcept {
    if (_immediateMode || !_workerAttached) return;

    EventSem event;
    EventSem *const pEvent = &event;

    ReaderContext *const pR = &_reader;

    ccnew_placement(allocate<TerminateConsumerThreadMessage>(1)) TerminateConsumerThreadMessage(pEvent, pR);

    kick();
    event.wait();

    // Join consumer thread to avoid MessageQueue instance is deleted before reading `_reader.terminateConsumerThread` and setting `_workerAttached`
    if (_consumerThread->joinable()) {
        _consumerThread->join();
    }
}
```
